# TRABAJO GRUPAL CORONAVIRUS

Group B: India, Peru, EEUU, Francia and Spain:

Kapil
María
Juan

---------------------

In [4]:
import pandas as pd
import numpy as np
import sys 
import os 
import json
from datetime import datetime

In [5]:
root_path = os.path.dirname(os.getcwd())
sys.path.append(root_path)
sys.path = list(set(sys.path)) 

In [6]:
import src.utils.visualization_tb as vis 
import src.utils.folders_tb as fol
import src.utils.mining_data_tb as mng
import src.utils.apis_tb as api

## 2.Data Wrangling

In [7]:
data = pd.read_csv("https://covid.ourworldindata.org/data/owid-covid-data.csv") 
data.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-02-24          1.0        1.0   
1      AFG      Asia  Afghanistan  2020-02-25          1.0        0.0   
2      AFG      Asia  Afghanistan  2020-02-26          1.0        0.0   
3      AFG      Asia  Afghanistan  2020-02-27          1.0        0.0   
4      AFG      Asia  Afghanistan  2020-02-28          1.0        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         NaN                  NaN  ...   
1                 NaN           NaN         NaN                  NaN  ...   
2                 NaN           NaN         NaN                  NaN  ...   
3                 NaN           NaN         NaN                  NaN  ...   
4                 NaN           NaN         NaN                  NaN  ...   

   gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
0        1803.987              NaN                597.029   
1        1803.987              NaN                597.029   
2        1803.987              NaN                597.029   
3        1803.987              NaN                597.029   
4        1803.987              NaN                597.029   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59             NaN           NaN                  37.746   
1                 9.59             NaN           NaN                  37.746   
2                 9.59             NaN           NaN                  37.746   
3                 9.59             NaN           NaN                  37.746   
4                 9.59             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  
0                         0.5            64.83                    0.498  
1                         0.5            64.83                    0.498  
2                         0.5            64.83                    0.498  
3                         0.5            64.83                    0.498  
4                         0.5            64.83                    0.498  

[5 rows x 55 columns]

In [8]:
data.shape

(61467, 55)

# 3. Data Cleaning

In [16]:
#Selecting assigned group B countries
data_sc = mng.country_select(df=data, count_list=['India', 'Peru', 'United States', 'France', 'Spain'])
data_sc.head()

iso_code continent location        date  total_cases  new_cases  \
24778      IND      Asia    India  2020-01-30          1.0        1.0   
24779      IND      Asia    India  2020-01-31          1.0        0.0   
24780      IND      Asia    India  2020-02-01          1.0        0.0   
24781      IND      Asia    India  2020-02-02          2.0        1.0   
24782      IND      Asia    India  2020-02-03          3.0        1.0   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
24778                 NaN           NaN         NaN                  NaN  ...   
24779                 NaN           NaN         NaN                  NaN  ...   
24780                 NaN           NaN         NaN                  NaN  ...   
24781                 NaN           NaN         NaN                  NaN  ...   
24782                 NaN           NaN         NaN                  NaN  ...   

       gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
24778        6426.674             21.2                 282.28   
24779        6426.674             21.2                 282.28   
24780        6426.674             21.2                 282.28   
24781        6426.674             21.2                 282.28   
24782        6426.674             21.2                 282.28   

       diabetes_prevalence  female_smokers  male_smokers  \
24778                10.39             1.9          20.6   
24779                10.39             1.9          20.6   
24780                10.39             1.9          20.6   
24781                10.39             1.9          20.6   
24782                10.39             1.9          20.6   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
24778                   59.55                        0.53            69.66   
24779                   59.55                        0.53            69.66   
24780                   59.55                        0.53            69.66   
24781                   59.55                        0.53            69.66   
24782                   59.55                        0.53            69.66   

       human_development_index  
24778                     0.64  
24779                     0.64  
24780                     0.64  
24781                     0.64  
24782                     0.64  

[5 rows x 55 columns]

In [14]:
#Percentage of missing values.  This will help us decide which columns we have selected.
miss = data_sc.isnull().sum() * 100 / len(data_sc)
miss.sort_values()

iso_code                                  0.000000
hospital_beds_per_thousand                0.000000
female_smokers                            0.000000
diabetes_prevalence                       0.000000
cardiovasc_death_rate                     0.000000
gdp_per_capita                            0.000000
aged_70_older                             0.000000
aged_65_older                             0.000000
median_age                                0.000000
population_density                        0.000000
population                                0.000000
life_expectancy                           0.000000
human_development_index                   0.000000
total_cases_per_million                   0.000000
continent                                 0.000000
location                                  0.000000
date                                      0.000000
total_cases                               0.000000
new_cases                                 0.056786
new_cases_per_million          

In [15]:
# Function used to see the different values inside each column. If there were only five values, that column was not included. 
data_sc["aged_65_older"].unique()

array([19.718,  5.989,  7.151, 19.436, 15.413])

In [19]:
#Selecting the columns to analyse.
data_final = data_sc[["iso_code","continent","location","date","total_cases","total_cases_per_million","new_cases_smoothed","new_cases_smoothed_per_million","new_deaths_smoothed","new_deaths_smoothed_per_million","total_deaths","total_deaths_per_million","life_expectancy","stringency_index","new_deaths"]]
data_final

iso_code continent location        date  total_cases  \
24778      IND      Asia    India  2020-01-30          1.0   
24779      IND      Asia    India  2020-01-31          1.0   
24780      IND      Asia    India  2020-02-01          1.0   
24781      IND      Asia    India  2020-02-02          2.0   
24782      IND      Asia    India  2020-02-03          3.0   
...        ...       ...      ...         ...          ...   
51838      ESP    Europe    Spain  2021-01-17    2252164.0   
51839      ESP    Europe    Spain  2021-01-18    2336451.0   
51840      ESP    Europe    Spain  2021-01-19    2370742.0   
51841      ESP    Europe    Spain  2021-01-20    2412318.0   
51842      ESP    Europe    Spain  2021-01-21    2456675.0   

       total_cases_per_million  new_cases_smoothed  \
24778                    0.001                 NaN   
24779                    0.001                 NaN   
24780                    0.001                 NaN   
24781                    0.001                 NaN   
24782                    0.002                 NaN   
...                        ...                 ...   
51838                48169.703           28829.143   
51839                49972.449           32095.571   
51840                50705.871           33360.286   
51841                51595.106           33747.000   
51842                52543.822           34958.286   

       new_cases_smoothed_per_million  new_deaths_smoothed  \
24778                             NaN                  NaN   
24779                             NaN                  NaN   
24780                             NaN                  NaN   
24781                             NaN                  NaN   
24782                             NaN                  NaN   
...                               ...                  ...   
51838                         616.603              205.714   
51839                         686.466              213.429   
51840                         713.516              212.857   
51841                         721.787              251.286   
51842                         747.694              280.286   

       new_deaths_smoothed_per_million  total_deaths  \
24778                              NaN           NaN   
24779                              NaN           NaN   
24780                              NaN           NaN   
24781                              NaN           NaN   
24782                              NaN           NaN   
...                                ...           ...   
51838                            4.400       53314.0   
51839                            4.565       53769.0   
51840                            4.553       54173.0   
51841                            5.375       54637.0   
51842                            5.995       55041.0   

       total_deaths_per_million  life_expectancy  stringency_index  new_deaths  
24778                       NaN            69.66             10.19         NaN  
24779                       NaN            69.66             10.19         NaN  
24780                       NaN            69.66             10.19         NaN  
24781                       NaN            69.66             10.19         NaN  
24782                       NaN            69.66             10.19         NaN  
...                         ...              ...               ...         ...  
51838                  1140.290            83.56             71.30         0.0  
51839                  1150.021            83.56             71.30       455.0  
51840                  1158.662            83.56             71.30       404.0  
51841                  1168.586            83.56             71.30       464.0  
51842                  1177.227            83.56               NaN       404.0  

[1766 rows x 15 columns]

In [20]:
#Changing date column to datatime
mng.changing_dates(df=data_final,name_column="date")  

In [21]:
#Information of our columns to see if changes of format are needed
data_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1766 entries, 24778 to 51842
Data columns (total 15 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         1766 non-null   object        
 1   continent                        1766 non-null   object        
 2   location                         1766 non-null   object        
 3   date                             1766 non-null   datetime64[ns]
 4   total_cases                      1766 non-null   float64       
 5   total_cases_per_million          1766 non-null   float64       
 6   new_cases_smoothed               1740 non-null   float64       
 7   new_cases_smoothed_per_million   1740 non-null   float64       
 8   new_deaths_smoothed              1740 non-null   float64       
 9   new_deaths_smoothed_per_million  1740 non-null   float64       
 10  total_deaths                     1620 non-null   float6

## 4.Creating an API.  

### We are creating the API in the file ../src/services/apis/server.py

4.1 - Create a dataframe with the mean new deaths of our selected countries.

4.2 - Transform this dataframe to a json format

4.3 - Create API with the json 

4.4 - Send the API to another group

### API workflow:

![API](../resources/API_info.png)